src/Workspaces/Core/Portable/Workspace/Solution/ProjectDependencyGraph.cs

```diff
@@ -14,35 +14,288 @@ namespace Microsoft.CodeAnalysis
     /// </summary>
     public class ProjectDependencyGraph
```

Comment:
>ProjectDependencyGraph [](start = 17, length = 22)

Would it make sense to make this `ImmutableDependencyGraph<T>`? Doesn't seem like anything is dependent on the nodes to be ProjectIds.

Reply:
It'd be easier to test. No need to create Solution, Projects, etc.

---
In reply to: [210755172](https://github.com/dotnet/roslyn/pull/28441#discussion_r210755172) [](ancestors = 210755172)

Reply:
@tmat: public API, so not changeable.

Reply:
Oh, didn't realized that.


---
In reply to: [210765503](https://github.com/dotnet/roslyn/pull/28441#discussion_r210765503) [](ancestors = 210765503)

Reply:
Note: even if ProjectDependencyGraph is public, there is some merit in the idea that it sits on top of an internal, more abstract/generic graph impl.  I am sympathetic to the idea of doing htat, especially as it would make things nicer from a testing perspective.  

That said, i don't think spending oodles of time here is very valuable.  So it's up to jason if there would be enough value in such a change.

Reply:
When we have the need for dependency graphs of other things, the refactoring would make great sense. In the mean time, customers are crashing due to OOMs. :smile: